File: Part5_Association_Rule_Learning/Section24_Apriori/apriori.py
# Apriori

# Data Preprocessing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
# from sklearn.cross_validation import train_test_split  # deprecated
from apyori import apriori


# Importing the dataset
os.chdir("C:/Development/Courses/Kirill Eremenko Data Science Courses/\
Machine_Learning_A-Z/Part5_AssociationRuleLearning/Section24_Apriori")
dataset = pd.read_csv("Market_Basket_Optimisation.csv", header=None)

# To see the full array, run the following
np.set_printoptions(threshold=np.nan)

# Need to create a list of lists that has the transactions
transactions = []
for i in range(0, 7501):
    transactions.append([str(dataset.values[i, j]) for j in range(0,20)])

# Training Apriori on the dataset
# Minimum Support: Products purchased at least 3 times a day; 7 days a week;
# 7500 total transactions
# (3*7)/7500=0.0028
rules = apriori(transactions, min_support=0.003, min_confidence=0.2,
                min_lift=3, min_length=2)

# Visualising the results
results = list(rules)
results[0:5]

File: Part5_Association_Rule_Learning/Section24_Apriori/apriori.R
# Apriori

# Data Preprocessing

# Importing the dataset
setwd("C:/Development/Courses/Kirill Eremenko Data Science Courses/Machine_Learning_A-Z/Part5_AssociationRuleLearning/Section24_Apriori")
dataset <- read.csv("Market_Basket_Optimisation.csv", header=FALSE)

# Create a sparse matrix with 120 columns (one for each item) and 0 or 1 in the records (1 if bought, 0 if not)
# install.packages("arules")
library(arules)
dataset <- read.transactions("Market_Basket_Optimisation.csv", sep=",", rm.duplicates=TRUE)
summary(dataset)

itemFrequencyPlot(dataset, topN=10)

# Training Apriori on the dataset
# Minimum Support: Products purchased at least 3 times a day; 7 days a week; 7500 total transactions
# (3*7)/7500=0.0028
rules <- apriori(data=dataset, parameter=list(support=0.003, confidence=0.8))  # 0 rules because confidence too high
rules <- apriori(data=dataset, parameter=list(support=0.003, confidence=0.4))
rules <- apriori(data=dataset, parameter=list(support=0.003, confidence=0.2))
# (4*7)/7500=0.00373333
rules <- apriori(data=dataset, parameter=list(support=0.004, confidence=0.2))

# Visualising the results
inspect(sort(rules, by="lift")[1:10])
